### ISSUE TITLE
kube-proxy 1.7.0: health check server bind to 127.0.0.1 despite set 0.0.0.0 via --healthz-bind-address 

### ISSUE THREAD

**Issue by r0bj** [kubernetes/kubernetes#48600]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:
kube-proxy DaemonSet:
```
apiVersion: extensions/v1beta1
kind: DaemonSet
metadata:
  labels:
    component: kube-proxy
    k8s-app: kube-proxy
    kubernetes.io/cluster-service: "true"
    name: kube-proxy
    tier: node
  name: kube-proxy
  namespace: kube-system
spec:
  updateStrategy:
    type: RollingUpdate
  template:
    metadata:
      labels:
        component: kube-proxy
        k8s-app: kube-proxy
        kubernetes.io/cluster-service: "true"
        name: kube-proxy
        tier: node
    spec:
      containers:
      - command:
        - kube-proxy
        - --kubeconfig=/var/lib/kube-proxy/kubeconfig
        - --proxy-mode=iptables
        - --healthz-bind-address=0.0.0.0
        - --v=2
        image: gcr.io/google_containers/kube-proxy-amd64:v1.7.0
        imagePullPolicy: IfNotPresent
        name: kube-proxy
        securityContext:
          privileged: true
        volumeMounts:
        - mountPath: /var/lib/kube-proxy
          name: kubeconfig
          readOnly: true
      dnsPolicy: ClusterFirst
      hostNetwork: true
      restartPolicy: Always
      terminationGracePeriodSeconds: 30
      volumes:
      - hostPath:
          path: /var/lib/kube-proxy
        name: kubeconfig
```

Note option `--healthz-bind-address=0.0.0.0`

Despite option `--healthz-bind-address=0.0.0.0` kube-proxy bind port 10249 to `127.0.0.1`:

```
# docker ps -qf "label=io.kubernetes.container.name=kube-proxy" | xargs docker inspect --format '{{.State.Pid}}'
3859
# nsenter -t 3859 -n netstat -pant |grep -i listen
tcp        0      0 0.0.0.0:179             0.0.0.0:*               LISTEN      2123/bird
tcp        0      0 127.0.0.1:8500          0.0.0.0:*               LISTEN      729/consul
tcp        0      0 127.0.0.1:53            0.0.0.0:*               LISTEN      905/dnsmasq
tcp        0      0 10.14.30.110:53         0.0.0.0:*               LISTEN      905/dnsmasq
tcp        0      0 10.14.30.110:22         0.0.0.0:*               LISTEN      786/sshd
tcp        0      0 127.0.0.1:8600          0.0.0.0:*               LISTEN      729/consul
tcp        0      0 10.14.30.110:5666       0.0.0.0:*               LISTEN      792/xinetd
tcp        0      0 127.0.0.1:10248         0.0.0.0:*               LISTEN      1307/kubelet
tcp        0      0 127.0.0.1:10249         0.0.0.0:*               LISTEN      3859/kube-proxy
tcp        0      0 10.14.30.110:8301       0.0.0.0:*               LISTEN      729/consul
tcp6       0      0 :::10256                :::*                    LISTEN      3859/kube-proxy
tcp6       0      0 ::1:53                  :::*                    LISTEN      905/dnsmasq
tcp6       0      0 :::4194                 :::*                    LISTEN      1307/kubelet
tcp6       0      0 :::10250                :::*                    LISTEN      1307/kubelet
tcp6       0      0 :::10255                :::*                    LISTEN      1307/kubelet
```

Note:
```
tcp        0      0 127.0.0.1:10249         0.0.0.0:*               LISTEN      3859/kube-proxy
```

**What you expected to happen**:

kube-proxy should bind port 10249 (health check server) to 0.0.0.0.

**Environment**:
- Kubernetes version (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.0", GitCommit:"d3ada0119e776222f11ec7945e6d860061339aad", GitTreeState:"clean", BuildDate:"2017-06-29T23:15:59Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.0", GitCommit:"d3ada0119e776222f11ec7945e6d860061339aad", GitTreeState:"clean", BuildDate:"2017-06-29T22:55:19Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
- OS (e.g. from /etc/os-release): Ubuntu 14.04
- Kernel (e.g. `uname -a`):
Linux dev-k8s-deployer-p1 4.4.0-34-generic #53~14.04.1-Ubuntu SMP Wed Jul 27 16:56:40 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

**Comment by cmluciano** [comment]
@kubernetes/sig-network-misc

**Comment by MrHohn** [comment]
Sorry for the confusion, I forgot to update the release note in #44968.

In v1.7, kube-proxy's healthz server will serve on [port 10256](https://github.com/kubernetes/kubernetes/blob/v1.7.0/pkg/master/ports/ports.go#L41-L43) instead, setting flag `healthz-bind-address` should affect this port. Meanwhile, [port 10249](https://github.com/kubernetes/kubernetes/blob/v1.7.0/pkg/master/ports/ports.go#L20-L22) will be mainly used for metrics server.

For transition, kube-proxy will continue serving `/healthz` on port 10249, but currently `MetricsBindAddress` is not adjustable through flag (I think we need a flag though).

**Comment by bowei** [comment]
/assign @MrHohn 

please close when the docs are updated

**Comment by jimmycuadra** [comment]
This broke our production deployment because our AWS ELBs were using port 10249 for health checks, and with this change kube-proxy is no longer listening for outside connections on this port. 😠

**Comment by MrHohn** [comment]
@jimmycuadra I'm terribly sorry for the production issue you encountered.

I'm cherrypicking #49799 and also patching in the release note for kube-proxy we should have published with the initial 1.7 release. Hopefully that could prevent more users being hurt by this and make the potential transition easier.

**Comment by jimmycuadra** [comment]
Thanks @MrHohn! I know it's hard to avoid breaking changes in software, especially a project as big and complicated as Kubernetes. To be clear, I place no blame on you or any other individual for the mistake. I do think it would be useful to loop in a project manager for a post mortem, cause I thiink it's important to do a post-mortem on issues like this to see how the process can be improved to detect and prevent similar issues in the future. The only manager on the project I know by name is @apsinha—Aparna, would you be so kind as to let an appropriate person know about this issue for the purposes of a post-mortem?
